packages/runtime/src/plugins/useDemandControl.ts

```diff
@@ -0,0 +1,102 @@
+
+export interface DemandControlPluginOptions {
+  /**
+   * The assumed maximum size of a list for fields that return lists.
+   */
+  defaultAssumedListSize?: number;
+  /**
+   * 	The maximum cost of an accepted operation. An operation with a higher cost than this is rejected.
+   *  If not provided, no maximum cost is enforced.
+   */
+  max?: number;
+  /**
+   * Whether to show the estimated cost in the extensions of the response.
+   * @default false
+   */
+  showInformationInExtensions?: boolean;
+}
+
+export function useDemandControl<TContext extends Record<string, any>>({
+  defaultAssumedListSize,
+  max,
+  showInformationInExtensions,
+}: DemandControlPluginOptions): GatewayPlugin<TContext> {
+  const calculateCost = createCalculateCost(defaultAssumedListSize);
+  const costByContextMap = new WeakMap<any, number>();
+  return {
+    onSubgraphExecute({ subgraph, executionRequest, logger }) {
+      const demandControlLogger = logger?.child('demand-control');
+      let costByContext = executionRequest.context
+        ? costByContextMap.get(executionRequest.context) || 0
+        : 0;
+      const operationCost = calculateCost(
+        subgraph,
+        executionRequest.document,
+        executionRequest.variables || EMPTY_OBJECT,
+      );
+      costByContext += operationCost;
+      if (executionRequest.context) {
+        costByContextMap.set(executionRequest.context, costByContext);
+      }
+      demandControlLogger?.debug({
+        operationCost,
+        totalCost: costByContext,
+      });
+      if (max != null && costByContext > max) {
+        throw createGraphQLError(
+          `Operation estimated cost ${costByContext} exceeded configured maximum ${max}`,
+          {
+            extensions: {
+              code: 'COST_ESTIMATED_TOO_EXPENSIVE',
+              cost: {
+                estimated: costByContext,
+                max,
+              },
+            },
+          },
+        );
+      }
+    },
+    onExecutionResult({ result, setResult, context }) {
+      if (showInformationInExtensions) {
+        const costByContext = costByContextMap.get(context);
+        if (costByContext) {
+          if (isAsyncIterable(result)) {
+            setResult(
+              mapAsyncIterator(result, (value) => ({
+                ...value,
+                extensions: {
+                  ...(value.extensions || {}),
+                  cost: {
+                    estimated: costByContext,
+                    max,
+                  },
+                },
+              })),
+            );
+          } else {
+            setResult({
+              ...(result || {}),
+              extensions: {
+                ...(result?.extensions || {}),
+                cost: {
+                  estimated: costByContext,
+                  max,
+                },
+              },
+            });
+          }
+          costByContextMap.delete(context);
```

Comment:
technically not necessary because of being a weakmap. But I am a fan of explicit maps... They tend to be slightly more performant.

Reply:
Actually I am going to remove that "delete" line. I prefer WeakMaps over touching the context. I am not sure Maps are safe because in case of an error it might cause a leak. And Maps are beneficial when you need keys if they are objects or keys are primitives, no? We usually use WeakMaps in other plugins to pass data between hooks. I just reused the same pattern here. Never heard about performance benefit of Maps over WeakMaps. Any resources about that?

Reply:
It's a micro-optimization but you can run a benchmark. I found a reasonable benchmark @ https://gist.github.com/jung-kim/d146b25ea754e73bf60a?permalink_comment_id=3811484#gistcomment-3811484